src/vm/ilstubresolver.cpp

```diff
@@ -318,42 +337,67 @@ ILStubResolver::AllocGeneratedIL(
 #if !defined(DACCESS_COMPILE)
     _ASSERTE(0 != cbCode);
 
-    NewArrayHolder<BYTE>             pNewILCodeBuffer        = NULL;
-    NewArrayHolder<BYTE>             pNewLocalSig            = NULL;
-    NewArrayHolder<CompileTimeState> pNewCompileTimeState    = NULL;
+    if (!UseLoaderHeap())
+    {
+        NewArrayHolder<BYTE>             pNewILCodeBuffer = NULL;
```

Comment:
Can you pull these command variables outside the branch. I don't see why they need to be different.

Reply:
The reason the branching here is weird is because we want to be able to use the holders for cleanup in the exceptional/failure case. However, when we don't need to use the loader heap, we allocate via "new" and delete in the `ClearCompileTimeState` method. If we use the loader heap, we want to allocate from the loader heap and use the appropriate holder and only delete when the loader is unloaded (not in `ClearCompileTimeState`). If we don't keep the data after `ClearCompileTimeState`, the JIT will assert when trying to JIT a P/Invoke or struct marshal stub that calls this struct marshal stub.

It's a really awkward case where RAII doesn't quite do what we want unless we create a "multi-holder" type from which we could pick which holder from a set we want to use.